lib/logmsg/nvtable.h

```diff
@@ -151,7 +168,25 @@ struct _NVEntry
       gchar data[];
     } vdirect;
 
-    NVReferencedSlice vindirect;
+    struct
+    {
+      NVHandle handle;
+      guint32 ofs;
+      guint32 len;
+
+      /* NOTE: this type field was promoted up into NVEntry, so we won't need
+       * the "type" field here, we are moving the value stored here in existing
+       * serialized messages in _update_entry() in logmsg-serialize-fixup.c.  We
+       * might be able to reuse this byte once we drop compatibility with
+       * version 26 of the LogMessage serialization format.
+       *
+       * NOTE: we zero out this field upon reading and then reserializing a message.
+       */
+
+      guint8 __deprecated_type_field;
+
+      gchar name[0];
```

Comment:
I was giving some thought to this when I made this changes, and I honestly think that it is not a good idea to couple the representation of the indirect entry with the arguments passed in a public interface.

The public interface should be tailored to the ease of use of its users, while the internal representation needs to match the on-disk format. These are or should be separate concerns and thus separate motivations for changing one or the other. This very example shows that a change in the on-disk format caused a cascaded change in the public interface of NVTable.

With this said, it is also true that I would prefer the non-struct based arguments to nv_table_add_value_indirect() feeling it is much easier to pass those arguments as separate parameters of the function instead of having to initialize a struct and pass that:

So this:

```
          success = nv_table_add_value_indirect(tab, handle, name, strlen(name),
                                                                         STATIC_HANDLE, 1, 126, 0, NULL);
```

instead of this:
```
          success = nv_table_add_value_indirect(tab, handle, name, strlen(name),
                                                &(NVReferencedSlice)
          {
            STATIC_HANDLE, 1, 126
          }, 0, NULL);
```

Also, the __deprecated_type_field would need to be zero initialized if we are to copy the struct contents as a whole, exactly because the on-disk format needs it to be zero-initialized, and pushing this burden to callers is not right, IMHO.

Honestly, I was originally planning to get rid of NVReferencedSlice also in the function prototype, but it was not strictly needed for this branch and I didn't want to do that since you so tiredlessly introduced it originally.

With all this said, I would either:
  1) leave NVReferencedSlice decoupled from NVEntry OR
  2) get rid off NVReferencedSlice completely, even from the function calls (thereby reverting one of your earlier patches almost completely).

I am willing to do the 2nd option if you agree with this explanation, but I am fine with the 1st, which is the current state in this branch.

Reply:
Okay, thank you for the explanation.

I'd vote for `1.`, because I would really like to keep `NVReferencedSlice`. In my opinion, `NVReferencedSlice` was half a step towards making the interface cleaner along clean-code principles by reducing a 9-parameter function to 6, grouping those 4 together and giving them a name. This idea could have been continued, but I stopped there.

On the user side, this looks like this (the compound literal looks terrible, I agree):
```
NVReferencedSlice referenced_slice =
  {
    .handle = ref_handle,
    .ofs = start_offset_in_refed_value,
    .len = length,
    .type = type
  };

nv_table_add_value_indirect(self->payload, ..., &referenced_slice, &new_entry));
```